### PR TITLE
[Bugfix]mask premature eos during in_reasoning for DeepSeek-V4

### LINKED ISSUE #14031: [RFC]: Phase-aware EOS policy for reasoning and tool-call generation
## Motivation

> **Status:** Future improvement and design discussion. This is not proposed as the immediate fix for #14013.

Reasoning and tool-calling models generate structured output in phases, for example:

```text
reasoning -> reasoning-end -> content or tool call -> EOS
```

Today, vLLM exposes request-wide controls such as `ignore_eos` and `min_tokens`, while reasoning and tool parsers mainly classify tokens after sampling. The existing `thinking_token_budget` machinery can force a reasoning-end sequence when a budget is exhausted, but there is no symmetric, phase-aware policy that can suppress EOS only while an explicitly protected reasoning phase is still open and restore EOS afterwards.

This leaves a gap for model families where EOS is not a valid terminator inside an established reasoning/tool-call protocol. A numerically plausible EOS can terminate generation before the reasoning-end or tool-call marker, producing `finish_reason="stop"` without a tool call.

The DeepSeek V4 investigation in #14013 exposed this as a separate follow-up surface:

- the original Ascend thinking workload produced `7/200` premature-EOS / stop-without-tool-call responses;
- two PA_ND ordering candidates removed physical-allocation-dependent operator output, but still produced `1/21` and `2/34` premature-EOS responses;
- the model-owner-confirmed matching DeepSeek API model completed `100/100` semantically equivalent requests with tool calls and no explicit EOS;
- a temperature-0 official API probe exposed processed top-logprobs for 121 reasoning tokens, and EOS was never returned with a finite logprob.

The official server rendered 56,894 prompt tokens versus 56,815 locally, and the API exposes processed top-logprobs rather than raw full-vocabulary logits. Therefore this evidence does **not** prove that the official implementation masks EOS, nor is it token-exact golden evidence. It does motivate discussing whether vLLM needs an explicit termination-policy abstraction for reasoning/tool protocols.

This RFC is not a proposed fix for the PA_ND allocation-order defect in #14013. Operator repeatability and generation termination policy are separate correctness surfaces.

Upstream context:

- https://github.com/vllm-project/vllm/issues/48645 tracks a DeepSeek V4 parser failure when generation reaches EOS without `</think>`. That is a parser/finalization issue; this RFC concerns whether sampling should permit EOS in an established protected phase.
- vLLM already has `ReasoningConfig`, reasoning start/end token IDs, `ThinkingBudgetStateHolder`, and logits-processor infrastructure. The proposal should extend those generic mechanisms instead of adding an Ascend-only sampler behavior.

## Proposed Change

Introduce an opt-in, phase-aware generation termination policy in upstream vLLM, then validate and integrate it in vLLM Ascend.

### Proposed semantics

Add a policy concept similar to:

```text
premature_eos_policy = "allow" | "mask_in_reasoning"
```

The exact public API is open for discussion. The required semantics are:

1. Default to `allow` for backward compatibility.
2. Derive phase state from actual prompt/generated token sequences and parser metadata, not only from a configured `thinking=true` flag.
3. With `mask_in_reasoning`, mask EOS only after an actual reasoning-start sequence has established an open reasoning phase.
4. Keep reasoning-end and valid tool-start sequences available while EOS is masked.
5. Restore EOS immediately after reasoning-end, an accepted implicit reasoning-end such as tool-start, or another model-declared transition.
6. Preserve `max_tokens` as a hard termination bound.
7. Apply the policy before sampling and top-logprob extraction so returned logprobs describe the effective distribution. Document that they are post-processor logprobs.
8. Define interactions with `min_tokens`, `ignore_eos`, structured output, speculative decoding/MTP, and multi-token phase delimiters.

The policy should be request- or model-configurable and remain opt-in until a model-family contract and regression evidence justify a default. It must not infer that every request with tools requires a tool call: `tool_choice="auto"` can legitimately return content or stop. A stricter tool-required policy, if desired, should be a separate explicit contract.

### Suggested implementation direction

Upstream vLLM:

- extend `ReasoningConfig` with termination-policy metadata;
- reuse or generalize the per-request state tracking in `ThinkingBudgetStateHolder`;
- apply EOS masking in the common logits-processing path for both normal and speculative sampling;
- expose a request-level override through OpenAI-compatible extra fields only after the semantics are agreed;
- keep parser finalization behavior, including https://github.com/vllm-project/vllm/issues/48645, as a separate fallback for outputs that still terminate with a half-open phase.

vLLM Ascend:

- consume the upstream generic policy rather than maintaining an Ascend-specific API;
- ensure the policy is applied consistently in eager/graph, synchronous/async, unified/PD, and MTP paths;
- add DeepSeek V4 A2/A3 regression coverage with BF16 and quantized checkpoints where practical.

### Alternatives considered

**Request-wide `ignore_eos=True`**

Too broad. It also suppresses valid EOS after reasoning/content completion and can force generation to `max_tokens`.

**A large `min_tokens` value**

Length-based rather than protocol-based. It can suppress valid short answers and still permit EOS before a late phase transition.

**Parser-only fallback**

Useful for routing or cleaning a response after termination, as discussed in https://github.com/vllm-project/vllm/issues/48645, but it cannot recover tokens that were never generated after premature EOS.

**Enable the policy automatically whenever thinking is configured**

Unsafe. Some models can answer directly without emitting an explicit reasoning-end marker. Automatic EOS suppression could turn a valid direct answer into a length-limited continuation. Phase establishment and model-family behavior must be explicit.

### Draft validation plan

Unit tests:

- EOS is highest before an established reasoning phase: unchanged under the default policy.
- EOS is highest inside an established protected phase: EOS is masked and a valid continuation/end token is selected.
- reasoning-end or tool-start transitions the request out of the protected phase and restores EOS.
- multi-token reasoning delimiters work across decode steps.
- mixed batches maintain independent phase state.
- returned top-logprobs reflect the post-policy distribution.
- `max_tokens`, `min_tokens`, `ignore_eos`, and structured-output precedence is explicit.

End-to-end tests:

- streaming and non-streaming reasoning plus tool calls;
- normal sampling and MTP/speculative decoding;
- eager and graph execution;
- synchronous and async scheduling;
- unified and PD-separated serving;
- DeepSeek V4 on A2/A3, with BF16 and quantized checkpoints where available;
- direct-answer cases without reasoning delimiters, to prove the opt-in policy does not regress parser fallback behavior.

Success metrics should keep deterministic repeatability separate from task behavior: phase-state correctness, premature-EOS count, tool-call validity, finish reason, output length, and latency overhead should all be reported independently.

## Feedback Period

One week: 2026-08-11 through 2026-08-18.

## Any Other Things

Questions for maintainers:

1. Should the generic abstraction live in `ReasoningConfig`, the logits-processor plugin interface, or a broader generation protocol/state-machine layer?
2. Should the first implementation support only reasoning EOS protection, or also explicit required-terminal policies for tool protocols?
3. Should request-level control use a vLLM extension field, or should it initially be server/model configuration only?
4. Which project should own the first implementation? The proposed preference is upstream vLLM for semantics and vLLM Ascend for backend integration and regression coverage.

### PR DESCRIPTION
**Pull request by doubinfx**
Fixes #14031 

### What this PR does / why we need it?

  This PR fixes the DeepSeek-V4 premature EOS issue on Ascend NPU during reasoning. When reasoning has started with `<think>` but has not reached `</think>`, EOS
  token id `1` can be sampled unexpectedly and terminate generation early.

  The fix is implemented in vLLM-Ascend only and is gated by an explicit Ascend config:

  {"premature_eos_policy": "mask_in_reasoning"}

  Scope is intentionally limited to the MVP requested:

  - DeepSeek-V4 only
  - EOS token id: 1
  - <think> token id: 128821
  - </think> token id: 128822
  - Uses SamplingParams.eos_token_id
  - Disabled when ignore_eos=True
  - Default behavior remains allow

  ### Does this PR introduce any user-facing change?

  Yes, but only when explicitly enabled.

  A new Ascend additional config option is added:

  {"premature_eos_policy": "mask_in_reasoning"}

  By default, premature_eos_policy is allow, so existing behavior is unchanged.

  ### How was this patch tested?

  Related unit tests:

  pytest -q tests/ut/worker/test_thinking_budget_state.py tests/ut/test_ascend_config.py

  Result:

  46 passed, 3 warnings, 10 subtests passed in 0.98s

  An Ascend NPU tensor-level smoke test was also run. This is not a full model inference test. It directly verifies that
  AscendThinkingBudgetStateHolder.apply_to_logits() masks EOS token id 1 on real NPU logits tensors while the request is inside <think>, and restores EOS after </think>.

  Environment:

  - Hardware: Ascend A3
  - CANN: 8.5.1
  - torch-npu: 2.10.0.post4
  - vLLM-Ascend source: current PR branch

  Command:

  ```bash
  PYTHONPATH=/tmp/codex_test_shims:/tmp/codex_vllm_58d:/mnt/share_space/codex_dir/vllm-ascend:$PYTHONPATH \
  ASCEND_VISIBLE_DEVICES=2 \
  OMP_PROC_BIND=false \
  OMP_NUM_THREADS=1 \
  PYTORCH_NPU_ALLOC_CONF=expandable_segments:True \
  python /tmp/test_ascend_eos_mask.py
  <details>
  <summary>Temporary test script: /tmp/test_ascend_eos_mask.py</summary>

  from types import SimpleNamespace

  import torch
  import torch_npu
  from vllm.v1.sample.logits_processor.interface import BatchUpdate

  from vllm_ascend.worker.thinking_budget_state import (
      DSV4_EOS_TOKEN_ID,
      DSV4_THINK_END_TOKEN_ID,
      DSV4_THINK_START_TOKEN_ID,
      PREMATURE_EOS_POLICY_MASK_IN_REASONING,
      AscendThinkingBudgetStateHolder,
  )

  VOCAB_SIZE = DSV4_THINK_END_TOKEN_ID + 8

  torch.npu.set_device(0)
  device = torch.device("npu:0")

  reasoning_config = SimpleNamespace(
      reasoning_start_token_ids=[DSV4_THINK_START_TOKEN_ID],
      reasoning_end_token_ids=[DSV4_THINK_END_TOKEN_ID],
  )

  sampling_params = SimpleNamespace(
      ignore_eos=False,
      eos_token_id=DSV4_EOS_TOKEN_ID,
      thinking_token_budget=None,
  )

  def make_holder(num_spec_tokens):
      return AscendThinkingBudgetStateHolder(
          reasoning_config,
          max_num_seqs=4,
          num_spec_tokens=num_spec_tokens,
          device=device,
          is_pin_memory=False,
          premature_eos_policy=PREMATURE_EOS_POLICY_MASK_IN_REASONING,
      )

  def sync(holder, output_token_ids):
      holder.sync_batch(
          BatchUpdate(
              batch_size=1,
              removed=[],
              added=[
                  (
                      0,
                      sampling_params,
                      [DSV4_THINK_START_TOKEN_ID],
                      output_token_ids,
                  )
              ],
              moved=[],
          )
      )

  normal_output = []
  normal_holder = make_holder(num_spec_tokens=0)
  sync(normal_holder, normal_output)

  normal_logits = torch.zeros((1, VOCAB_SIZE), device=device)
  normal_holder.apply_to_logits(
      normal_logits,
      predict_bonus_token=False,
      spec_token_ids=None,
  )
  torch.npu.synchronize()
  normal_eos = normal_logits[0, DSV4_EOS_TOKEN_ID].cpu()
  print("NPU_NORMAL_EOS", normal_eos.item())
  assert torch.isneginf(normal_eos)

  normal_output.append(DSV4_THINK_END_TOKEN_ID)
  normal_holder.update_state([normal_output], spec_token_ids=None)

  after_logits = torch.zeros((1, VOCAB_SIZE), device=device)
  normal_holder.apply_to_logits(
      after_logits,
      predict_bonus_token=False,
      spec_token_ids=None,
  )
  torch.npu.synchronize()
  after_eos = after_logits[0, DSV4_EOS_TOKEN_ID].cpu().item()
  print("NPU_AFTER_END_EOS", after_eos)
  assert after_eos == 0.0

  spec_holder = make_holder(num_spec_tokens=2)
  sync(spec_holder, [])

  spec_logits = torch.zeros((2, VOCAB_SIZE), device=device)
  spec_holder.apply_to_logits(
      spec_logits,
      predict_bonus_token=False,
      spec_token_ids=[[10, DSV4_THINK_END_TOKEN_ID]],
  )
  torch.npu.synchronize()
  spec_values = [spec_logits[i, DSV4_EOS_TOKEN_ID].cpu() for i in range(2)]
  print("NPU_SPEC_EOS", [v.item() for v in spec_values])
  assert torch.isneginf(spec_values[0])
  assert torch.isneginf(spec_values[1])

  bonus_holder = make_holder(num_spec_tokens=2)
  sync(bonus_holder, [])

  bonus_logits = torch.zeros((1, VOCAB_SIZE), device=device)
  bonus_holder.apply_to_logits(
      bonus_logits,
      predict_bonus_token=True,
      spec_token_ids=[[10, DSV4_THINK_END_TOKEN_ID]],
  )
  torch.npu.synchronize()
  bonus_eos = bonus_logits[0, DSV4_EOS_TOKEN_ID].cpu().item()
  print("NPU_BONUS_AFTER_DRAFT_END_EOS", bonus_eos)
  assert bonus_eos == 0.0

  print("NPU_MASK_SMOKE_DONE")
  </details>
  ```

  Result:

  NPU_NORMAL_EOS -inf
  NPU_AFTER_END_EOS 0.0
  NPU_SPEC_EOS [-inf, -inf]
  NPU_BONUS_AFTER_DRAFT_END_EOS 0.0
  NPU_MASK_SMOKE_DONE

  Result interpretation:

  - NPU_NORMAL_EOS -inf: EOS token id 1 was masked while still inside <think>.
  - NPU_AFTER_END_EOS 0.0: EOS token id 1 was restored after </think>.
  - NPU_SPEC_EOS [-inf, -inf]: EOS was masked for speculative target logits while still inside reasoning.
  - NPU_BONUS_AFTER_DRAFT_END_EOS 0.0: EOS was restored for the bonus token when draft tokens already included </think>.
  - NPU_MASK_SMOKE_DONE: All assertions passed.

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

File: vllm_ascend/worker/thinking_budget_state.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        self,
+        output_token_ids: list[list[int]],
+        spec_token_ids: list[list[int]] | None,
+        repeat_indices: torch.Tensor | None = None,
+    ) -> None:
+        super().update_state(output_token_ids, spec_token_ids, repeat_indices)
+        if not self.is_enabled or not self._state:
+            return
+        for state in self._state.values():
+            self._update_eos_mask_state(state)
```
[the remark this answers is not in the document]
Comment (doubinfx) [reply vllm_ascend/worker/thinking_budget_state.py:102]: cu_num_tokens is a dict[int, int] in the current upstream holder, so if seq_idx not in self.cu_num_tokens is the correct key check.

File: vllm_ascend/worker/thinking_budget_state.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        logits = super().apply_to_logits(logits, predict_bonus_token, spec_token_ids)
+        if not self.is_enabled or not self._state or not self._mask_premature_eos:
+            return logits
+
+        spec_lists = spec_token_ids or []
+        eos_mask_indices_cpu: list[int] = []
+        eos_tokens_cpu: list[int] = []
+        for seq_idx in sorted(self._state.keys()):
+            if seq_idx not in self.cu_num_tokens:
+                continue
```
[the remark this answers is not in the document]
Comment (doubinfx) [reply vllm_ascend/worker/thinking_budget_state.py:119]: The upstream state key is prompt_tok_ids, not prompt_token_ids, so the current lookup matches the actual ThinkingBudgetStateHolder implementation.

File: vllm_ascend/worker/thinking_budget_state.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        return in_reasoning
+
+    def _advance_reasoning_state(self, in_reasoning: bool, token_ids: list[int]) -> bool:
+        for token_id in token_ids:
+            in_reasoning = self._consume_reasoning_token(in_reasoning, token_id)
+        return in_reasoning
+
+    def _init_eos_mask_state(self, state: dict[str, Any]) -> None:
+        prompt_tok_ids = state.get("prompt_tok_ids") or []
+        output_tok_ids = state.get("output_tok_ids") or []
```
[the remark this answers is not in the document]
Comment (doubinfx) [reply vllm_ascend/worker/thinking_budget_state.py:157]: super().update_state() already refreshes state["output_tok_ids"], including the repeat_indices path, so duplicating it here is unnecessary and could regress that layout.
